### ISSUE TITLE
Implement Future blocking adapters for blocking code

### ISSUE THREAD

**Issue by Eliah-Lakhin** [tokio-rs/tokio#1479]
## Motivation

Implementing intercommunication between the blocking and non-blocking code is still hard to achieve in Tokio `0.2` when we need to run continuous (potentially) blocking process.

An example of such process could be, in example, communicating with hardware drivers(i.e. OpenGL interface) that blocks OS thread in some particular cases for short time, but requires to be run continuously in order to perform it's work efficiently. Let's assume a use case when we need to implement a Network Multimedia Service that will do a video-decoding by demand. The video-decoding process will be implemented by GPU cluster in separated blocking threads, and the Network Clients will connect to our server using Tokio non-blocking infrastructure.

To organize intercommunication between blocking and non-blocking code we can use `tokio::sync::mpsc` channel to notify blocking code to schedule another bunch of work on GPU, and to use `tokio::sync::oneshot` to return result from blocking code to the Tokio non-blocking task.

The second one(blocking -> non-blocking through `oneshot`) is easy to achieve using existing Tokio API. But there is no idiomatic way to read messages from `mpsc::Receiver` channel end in the blocking code. One of the workaround could be instantiating a Tokio `current_thread` Executor inside the blocking code(instantiated with `tokio_threadpool::blocking`), and to block the blocking thread on `mpsc::Receiver` stream. But the current implementation of Tokio `0.2` doesn't provide a way to block the thread on futures(in opposite to version `0.1` that had a `Future::wait()` function). Instead we have to implement blocking manually using Executor methods, like: `tokio_executor::enter().unwrap().block_on(my_future)`.

To sum up, described goal is achievable through the Tokio `0.2`, but the solution is not clean enough from the API design point of view and requires more boilerplate code in contrast to Tokio `0.1` .

## Proposal

Implement blocking adapters for all `tokio::sync` prototypes to block the thread on Future execution similar to `Future::wait()` from Tokio `0.1`.

**Comment by sfackler** [comment]
You just need to call https://docs.rs/futures-preview/0.3.0-alpha.18/futures/executor/fn.block_on.html.

**Comment by Eliah-Lakhin** [comment]
@sfackler Well, it is achievable this way too, but still looks too verbose, isn't it?

Also, what are the functional differences between `futures::executor` from the Futures `0.3` and Tokio `current_thread`?

**Comment by sfackler** [comment]
The `current_thread` executor has the ability to run multiple spawned futures, while `executor::block_on` internally uses a very minimal executor that's just able to wait on a future to complete.

**Comment by Darksonn** [comment]
Closing as duplicate of #2629.
